include/lbann/callbacks/callback.hpp

```diff
@@ -89,6 +90,10 @@ class callback_base {
   /** @name Callback hooks */
   ///@{
 
+  /** @brief Called at the beginning of setup. */
+  virtual void on_setup_begin(model *m) {}
```

Comment:
This PR is not the right place to make this change, but these should all take `model&`. And they should all be `const`.

Reply:
It would be nicer if the callbacks worked on references. They shouldn't be `const` though, since the callback might manipulate the model or its members.

Reply:
The callback functions should be `const`:
```cpp
virtual void on_setup_begin(model&) const;
```

Reply:
I'm not sure about that either, since the callback might maintain some state or preallocate buffers.

Reply:
Invoking a callback should not affect it's *logical* state. If the bitwise state needs to change, the affected parameters should be marked `mutable` (and suitably protected with mutexes or atomics, etc, if multithreaded access is a concern).

Reply:
It looks like this is a question about whether callbacks should be considered pure functions or not. Making them pure functions will make them easier to reason about, but it also excludes some useful functionality. For example, a callback for learning rate schedules needs to store the initial learning rate (logically at `on_setup_end`). A callback that checkpoints the model at regular time intervals needs to keep track of the last checkpoint time.

Reply:
I have no objection to them having state -- they probably should. Memoization and/or other caching does not affect the fact that a callback is logically `const` whilst it's being applied -- nothing that defines the external character of the callback is changing (that is, if a change of state did occur, it would not be externally visible). For the learning rate schedule example, the initial learning rate is merely memoized. Neither the schedule or the mechanism for applying that schedule are affected by actually applying it. For the checkpoint interval, its progress is, again, internal cached state -- nothing externally visible occurs when the interval tracker is updated.

I vote we move this to another forum, perhaps an issue discussing the design of callbacks. As we've discussed previously, I think we need to rethink the interface (or eliminate it in favor of a generalized functor) and allow a wider variety of functors to be usable as callbacks. A few notable issues:
1. `const`-correctness is lacking (i.e., this discussion)
2. It's difficult to register new hooks and requires changing individual callbacks to allow them to work with new hooks.
3. The hooks that are there are clunky because they incur an overhead with `n` virtual function dispatches every time, even though many of those dispatches could be empty functions.
4. The interface belies the semantics of callbacks. That is, each hook takes (at least) a model as a parameter, implying that the callback should be applicable to whatever model gets passed into it. For example:
```cpp
model a, b;
callback::whatever cb;
cb->on_epoch_end(a);
cb->on_epoch_end(b);
```
There's no reason this code won't execute -- none of the callbacks explicitly register the model with which they were setup. There are no comments to suggest one should not execute this code. There's also no guarantee that any invariants of `callback::whatever` will be left intact after this code. For the examples of the callbacks that cache information about model state, for example, that information is probably invalid if a different model gets passed into the same callback object. Of course, this doesn't happen in the blessed executables of today because callbacks are only executed inside the model. However, neither are they implementation details of the model nor is their interface restricted away from users, opening them up to potential abuse of this kind as others build custom training algorithms or the like with the library components of LBANN.